app/src/main/java/com/rkm/tasky/repository/implementation/TaskyTaskRepositoryImpl.kt

```diff
@@ -0,0 +1,82 @@
+package com.rkm.tasky.repository.implementation
+
+import com.rkm.tasky.database.dao.SyncDao
+import com.rkm.tasky.database.dao.TaskDao
+import com.rkm.tasky.database.model.SyncEntity
+import com.rkm.tasky.database.model.SyncItemType
+import com.rkm.tasky.database.model.SyncUserAction
+import com.rkm.tasky.di.IoDispatcher
+import com.rkm.tasky.feature.task.model.Task
+import com.rkm.tasky.network.datasource.TaskyTaskRemoteDataSource
+import com.rkm.tasky.network.util.NetworkError
+import com.rkm.tasky.network.util.safeCall
+import com.rkm.tasky.repository.abstraction.TaskyTaskRepository
+import com.rkm.tasky.repository.mapper.asTask
+import com.rkm.tasky.repository.mapper.asTaskEntity
+import com.rkm.tasky.repository.mapper.asTaskyRequest
+import com.rkm.tasky.util.result.EmptyResult
+import com.rkm.tasky.util.result.Result
+import com.rkm.tasky.util.result.onFailure
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class TaskyTaskRepositoryImpl @Inject constructor(
+    private val remoteDataSource: TaskyTaskRemoteDataSource,
+    private val localDataSource: TaskDao,
+    private val syncDataSource: SyncDao,
+    @IoDispatcher private val  dispatcher: CoroutineDispatcher
+): TaskyTaskRepository {
+    override suspend fun getTask(id: String): Result<Task?, NetworkError.APIError> = withContext(dispatcher){
+        val results = safeCall { remoteDataSource.getTask(id) }
+        results.onFailure {
+            return@withContext Result.Success(localDataSource.getTaskById(id)?.asTask())
+        }
+        return@withContext Result.Success((results as Result.Success).data.asTask())
+    }
+
+    override suspend fun createTask(task: Task): EmptyResult<NetworkError.APIError> {
+        val results = safeCall { remoteDataSource.createTask(task.asTaskyRequest()) }
+        results.onFailure {
+            syncDataSource.upsertSyncItem(
+                SyncEntity(
+                    action = SyncUserAction.CREATE,
+                    item = SyncItemType.TASK,
+                    itemId = task.id
+                )
+            )
+        }
```

Comment:
To ensure that your sync call does not get cancelled before finishing execution (which can happen when the parent coroutine, like `viewModelScope`, is cancelled when the user navigates back), I’d wrap this in a scope that has a longer lifecycle. This is an unlikely situation, but it still should be accounted for.

The best way to do this is with a custom application wide coroutine scope, which lives as long as your app's process:
```
class MyApp: Application() {
    val applicationScope = CoroutineScope(SupervisorJob())
    
    // ...
}
```
You can then provide an instance of this scope in your Dagger module:
```
fun provideApplicationScope(app: Application): CoroutineScope {
    return (app as MyApp).applicationScope
}
```
Then, inject it into the class you need it in:
```
class MyRepository(
    private val applicationScope: CoroutineScope,
    ...
)
```
And make sure the second suspend function is executed in this new scope which won't be cancelled:
```
applicationScope.launch {
    // schedule sync
}.join()
```
This way, if the coroutine reached this point and wasn't cancelled, we make sure that it also won't be cancelled during this second call. The `join()` makes sure to still suspend the outer suspend function until the schedule sync call finished.